### PR TITLE
CMR-9889: Adding in a new index to make it faster to delete granules and to not stop the job if one provider's granule deletion fails.

### PR DESCRIPTION
**Pull request by eereiter**
# Overview

### What is the feature/fix?

Adding in a new index to make it faster to delete granules and to not stop the job if one provider's granule deletion fails.

### What is the Solution?

Adding in a new index to make it faster to delete granules and to not stop the job if one provider's granule deletion fails.

### What areas of the application does this impact?

job of deleting expired granules.

# Checklist

- [ ] I have updated/added unit and integration tests that prove my fix is effective or that my feature works
- [ ] New and existing unit and int tests pass locally and remotely with my changes
- [x] I have removed unnecessary/dead code and imports in files I have changed
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have cleaned up integration tests by doing one or more of the following:
  - de-duped, consolidated, removed dead int tests
  - transformed applicable int tests into unit tests
  - refactored to reduce number of system state resets by updating fixtures (use-fixtures :each (ingest/reset-fixture {})) to be :once instead of :each

### CODE REVIEW

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj [hunk 1]
```diff
@@ -4,13 +4,7 @@
...
    [cmr.metadata-db.data.util :as util]))
 
-(defmulti granule-column-sql
-  "Returns the sql to define provider granule columns"
-  (fn [provider]
-    (:small provider)))
-
-(defmethod granule-column-sql false
-  [provider]
+(def granule-column-sql-false
```
Comment (TylerHeald1) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: can this be changed to 'granule-column-sql-common' since it gets used for both small and non-small provider

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj [hunk 2]
```diff
@@ -30,20 +24,16 @@
...
-  (fn [provider table-name]
-    (:small provider)))
+  (case (:small provider)
+    false granule-column-sql-false
+    true (str granule-column-sql-false ",provider_id VARCHAR(255) NOT NULL")))
 
-(defmethod granule-constraint-sql false
-  [provider table-name]
-  (util/validate-table-name table-name)
+(defn- granule-constraint-sql-false
```
Comment (TylerHeald1) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:34]: Is there a name without a boolean value in it that would work well?

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj [hunk 3]
```diff
@@ -65,8 +55,9 @@
           table-name
           table-name))
 
-(defmethod granule-constraint-sql true
-  [provider table-name]
+(defn- granule-constraint-sql-true
```
Comment (TylerHeald1) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:58]: Is there a name without a boolean value in it that would work well?

**Review by TylerHeald1 — APPROVED** [review]

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj [hunk 4]
```diff
@@ -4,13 +4,7 @@
    [clojure.java.jdbc :as jdbc]
    [cmr.metadata-db.data.util :as util]))
 
-(defmulti granule-column-sql
```
Comment (jmaeng72) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: This is just a curiosity question. But are we trying to remove defmulti because it's hard to follow or not recommended design?

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj [hunk 5]
```diff
@@ -30,20 +24,16 @@
         transaction_id INTEGER DEFAULT 0 NOT NULL,
         created_at TIMESTAMP WITH TIME ZONE DEFAULT SYSTIMESTAMP NOT NULL"))
 
-(defmethod granule-column-sql true
+(defn granule-column-sql
```
Comment (jmaeng72) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:27]: For easier readability, I would like to suggest we start naming our functions with more active verbs. 

'get-granule-columns-sql-query'

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 2 above)
Comment (jmaeng72) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:34]: Suggestions:
- create-granule-constraint-sql-non-small-prov
- get-granule-constraint-sql-non-small-prov

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 5 above)
Comment (jmaeng72) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:27]: create-granule-columns-sql

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj [hunk 6]
```diff
@@ -89,6 +80,14 @@
           table-name
           table-name))
 
+(defn granule-constraint-sql
+  "Returns the sql to define constraint on provider granule table"
+  [provider table-name]
+  (util/validate-table-name table-name)
+  (case (:small provider)
+    false (granule-constraint-sql-false table-name)
```
Comment (jmaeng72) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:88]: Is it worth separating these into separate funcs? Why not keep the case statement here along with the sql query creation logic?

File: metadata-db-app/src/cmr/metadata_db/migrations/088_add_meta_to_provider.clj [hunk 7]
```diff
@@ -19,4 +19,4 @@
   []
   ;; do nothing as we don't want to loose the metadata documents when migrating
```
Comment (johnwteague) [inline metadata-db-app/src/cmr/metadata_db/migrations/088_add_meta_to_provider.clj:20]: nit: while you're here, can you fix the typo "loose"?

File: metadata-db-app/src/cmr/metadata_db/migrations/089_add_granule_delete_index.clj [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  (:require
+   [config.mdb-migrate-helper :as mh]))
+
+(defn up
+  "Migrates the database up to version 89."
+  []
+  (println "cmr.metadata-db.migrations.089-add-granule-delete-index up...")
+  (doseq [table-name (mh/get-concept-tablenames :granule)]
+    (try
+      (mh/sql (format "create index %s_DD ON %s (deleted, delete_time)" table-name table-name))
```
Comment (johnwteague) [inline metadata-db-app/src/cmr/metadata_db/migrations/089_add_granule_delete_index.clj:11]: don't you need a different index for small_providers, one that includes the provider_id?

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj [hunk 9]
```diff
@@ -107,32 +106,24 @@
...
   ;; This index makes it faster to find counts by collection of undeleted granules for metadata db holdings
-  (jdbc/db-do-commands
-    db (format "create index %s_pdcr on %s (parent_collection_id, deleted, concept_id, revision_id)"
-               table-name
-               table-name)))
+  (jdbc/db-do-commands db (format "create index %s_pdcr on %s (parent_collection_id, deleted, concept_id, revision_id)"
+                                  table-name
+                                  table-name))
+  ;; This index makes it much faster to find granules that are not deleted that have a delete time.
+  (jdbc/db-do-commands db (format "create index %s_dd on %s (deleted, delete_time)"
```
Comment (johnwteague) [inline metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:118]: Don't you need a different index for small_providers, one that includes the provider_id?

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 4 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: I read a while ago online - I can't find it now though, as well as from a previous CMR clojure developer who really knew clojure - that defmulti is slow processing wise.  I can't quite remember the numbers but it is something like 20 to 100 times slower than using other branching mechanisms. We really don't need polymorphism in these instances and in a lot of cases, I seem to be able to reduce code.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 1 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: Yes, the code has been updated.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 5 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:27]: I have updated the function names and used create-...

Comment (zimzoom) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:27]: +1

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 1 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: The function is now called create-granule-column-sql

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 2 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:34]: Yes, I changed the functions to use create-...   and nothing for the normal providers and I have added 
create-...-small-provider for the functions that represent the small provider.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 3 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:58]: yes this function is now called create-granule-constraint-sql since this function is for the normal providers.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 4 above)
Comment (jmaeng72) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: 20-100x slower?! wow.

Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: I can't find the exact number but when I read it, it was significant.  Since I can't find that article anymore, I also can't tell you how long ago it was written.  It is possible that clojure has fixed up the slowless using multidef, but not sure.  I could create a test and see, but I am finding that the code becomes cleaner when I don't use multidef  (and when I take the time to properly refactor the defmethods).

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 1 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: I refactored the main function and that function no longer exists.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 3 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:58]: I refactored the calling function and this function no longer exists.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 6 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:88]: No, I refactored this function and it no longer calls other functions.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 9 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:118]: Yes, Thank you for catching this!  I have fixed the mistake.

File: metadata-db-app/src/cmr/metadata_db/migrations/088_add_meta_to_provider.clj (hunk 7 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/migrations/088_add_meta_to_provider.clj:20]: fixed.

File: metadata-db-app/src/cmr/metadata_db/migrations/089_add_granule_delete_index.clj (hunk 8 above)
Comment (eereiter) [reply metadata-db-app/src/cmr/metadata_db/migrations/089_add_granule_delete_index.clj:11]: Yes, Thank you!  I have fixed the mistake.

File: metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj (hunk 4 above)
Comment (jmaeng72) [reply metadata-db-app/src/cmr/metadata_db/data/oracle/granule_table.clj:7]: I personally don't like multidef because it is difficult to follow, so I am onboard with this change

**Review by jmaeng72 — APPROVED** [review]

**Review by johnwteague — APPROVED** [review]
